### ISSUE TITLE
Referencing uninitialized data

### ISSUE THREAD

**Issue by dojt** [QuEraComputing/Bloqade.jl#464]
Hey guys,

Is it possible that `vm[m+1]` may be read out without having been initialized in the following line?

https://github.com/QuEraComputing/Bloqade.jl/blob/4c3b0be48a9840e6f5dd2abd307e44bc2c964717/lib/BloqadeKrylov/src/expmv.jl#L129

It seems to me that it could happen if the "happy breakdown" occurs for `j==m` in the 1st iteration of the while loop, no?

**Comment by dojt** [comment]
I modified the file `expmv.jl` in the `test`-folder by replacing `Float64` matrices and vectors with `Real` matrices and vectors:

```julia
@testset "expmv" begin
    A = Real[ rand() for k=1:10, ℓ=1:10 ]
    st = Real[ rand() for k=1:10 ]
    @test BloqadeKrylov.expmv(-0.1im, A, st) ≈ exp(-0.1im * A) * st

    _A = sprand(ComplexF64, 10, 10, 0.8)
    A = Complex{Real}[ _A[k,ℓ] for k=1:10, ℓ=1:10 ]
    st = Complex{Real}[ rand(ComplexF64) for k=1:10 ]
    @test BloqadeKrylov.expmv(-0.1im, A, st) ≈ exp(-0.1im * Matrix(A)) * st
end
```

The exception *`UndefRefError`: access to undefined reference* is then thrown all the time from that line... 🤔🤔🤔

**Comment by Roger-luo** [comment]
Good catch... but I'd say this is quite an edge case for evaluation of boxed array, could you elaborate on why this needs to be a `Real`? 

I think a simple fix could be to initialize them to zero since the actual value will be assigned later or not used anyways.

**Comment by dojt** [comment]
Using Real makes Julia throw an error for an access to memory that's undefined -- using Float just returns the uninitialized memory. In other words, with Float, the function computes with uninitialized memory.

**Comment by Roger-luo** [comment]
Right, I think using uninitialized memory here is fine, as the `Vector{Float64}(undef, n)` is usually small (tho not guaranteed), and thus `err1 >> err2` is always true. So we could change the initialization from `similar` to `zero` which will be a bit costly, but probably not a big deal at all.

This is why I ask what would be the use case for `Real` or the above case since I think this is really an edge case, it'd be nice to make a decision if there is an actual use case effected by this.

**Comment by dojt** [comment]
I disagree. Use of uninitialized memory in calculations is never fine -- even if *usually* it doesn't lead to problems. That is so even if it weren't the case that the function is generically typed, so that there's no way to predict what effects uninitialized reads have for the types it will be used with.
